### PR TITLE
Total websites count and list of websites

### LINKED ISSUE #11: [API] Websites List Endpoint for Hub Admins
* As a Protected API Client, I want to get a list of all unique websites that asked for notifications

### LINKED ISSUE #12: [API] Websites Count endpoint for Hub Admins
* As a Protected API Client, I want to get statistics overview with total websites count

### PR DESCRIPTION
**Pull request by tuka217**
Q	|A
-----|-----
Bug fix?|	no
New feature?|	yes
BC breaks?|	no
Fixed tickets|	resolve #11, #12 
License|	??

### CODE REVIEW

File: src/Sylius/Hub/Controller/HubController.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class HubController extends Controller
+{
+    public function statisticsAction(Request $request)
+    {
+        $entityManager = $this->get('doctrine.orm.entity_manager');
```
Comment (NeverResponse) [inline src/Sylius/Hub/Controller/HubController.php]: Wouldn't it be better to inject the repo directly instead of fetching it from the container on each call?
Would simplify things a bit, remove the extend, also give us more clarity as to what is actually being used inside.

File: src/Sylius/Hub/Repository/WebsiteRepository.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function countWebsites()
+    {
+        return (int) $this->createQueryBuilder('o')
+            ->select('count(o.id)')
+            ->getQuery()
+            ->getSingleScalarResult();
```
Comment (NeverResponse) [inline src/Sylius/Hub/Repository/WebsiteRepository.php]: `;` to the next line 💃

File: src/Sylius/Hub/Security/InMemoryApiUserProviderInterface.php [hunk 3]
```diff
@@ -24,11 +24,11 @@
     /**
      * @param string $apiKey
      *
-     * @return string
+     * @return UserInterface
      *
      * @throws UsernameNotFoundException
```
Comment (NeverResponse) [inline src/Sylius/Hub/Security/InMemoryApiUserProviderInterface.php]: `UserForTokenNotFoundException`

File: tests/Sylius/Hub/Controller/HubApiTest.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
+<?php
+
+use Lakion\ApiTestCase\JsonApiTestCase;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class HubApiTest extends JsonApiTestCase
```
Comment (NeverResponse) [inline tests/Sylius/Hub/Controller/HubApiTest.php]: final

File: src/Sylius/Hub/Controller/HubController.php (hunk 1 above)
Comment (tuka217) [inline src/Sylius/Hub/Controller/HubController.php]: Yea, but then you need to register repository as a service or use expression as an argument while registering the controller as a service. It should work, but doesn't. ^^
http://www.zuellich.de/blog/2016/03/symfony-3-inject-entity-repository-into-service-controller.html

File: spec/Sylius/Hub/Security/TokenAuthenticatorSpec.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    function it_return_json_response_with_information_that_authentication_is_required(Request $request)
+    {
+        $this->start($request)->shouldReturnAnInstanceOf(JsonResponse::class);
+    }
+
+    function it_returns_authentication_credentials_from_request(Request $request, HeaderBag $headers)
+    {
+        $request->headers = $headers;
```
Comment (pamil) [inline spec/Sylius/Hub/Security/TokenAuthenticatorSpec.php]: Anyway, according to "don't mock what you don't own" rule I think it would be better to create a real Symfony request here.

```
$request = new Request();
$request->headers->set('Authorization', 'SampleToken');
```

WDYT?

File: spec/Sylius/Hub/Security/TokenAuthenticatorSpec.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    function it_returns_nothing_if_authorization_credentials_were_not_passed(Request $request, HeaderBag $headers)
+    {
+        $request->headers = $headers;
+        $headers->get('Authorization')->willReturn(null);
+
+        $this->getCredentials($request)->shouldReturn(null);
+    }
+
+    function it_returns_user_for_given_token(InMemoryApiUserProviderInterface $apiUserProvider, UserInterface $user)
```
Comment (pamil) [inline spec/Sylius/Hub/Security/TokenAuthenticatorSpec.php]: `InMemoryApiUserProviderInterface` feels wrong, interface name should not determine its implementation, it probably should be renamed to `ApiUserProviderInterface`

File: src/Sylius/Hub/Controller/PackagistController.php [hunk 7]
```diff
@@ -1,37 +1,38 @@
...
+    /**
      * @return JsonResponse
      *
      * @throws HttpException
      */
-    public function getCurrentSyliusVersionAction(Request $request)
+    public function getCurrentSyliusVersionAction()
     {
-        $currentSyliusVersion = $this->get('hub.cached_sylius_version_provider')->provide();
+        $currentSyliusVersion = $this->syliusVersionProvider->provide();
```
Comment (pamil) [inline src/Sylius/Hub/Controller/PackagistController.php]: 👍

File: src/Sylius/Hub/Controller/StatisticsController.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class StatisticsController
+{
+    /**
+     * @var WebsiteRepositoryInterface
+     */
+    private $websiteRepository;
+
+    public function __construct(WebsiteRepositoryInterface $websiteRepository)
```
Comment (pamil) [inline src/Sylius/Hub/Controller/StatisticsController.php]: Docblock missing 🌮

File: src/Sylius/Hub/Controller/WebsitesController.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+     * @return Response
+     */
+    public function indexAction()
+    {
+        $websites = $this->websiteRepository->findAll();
+        $data = ['websites' => $websites];
+
+        $jsonWebsites = $this->serializer->serialize($data, 'json');
+
+        $response = new Response($jsonWebsites);
```
Comment (pamil) [inline src/Sylius/Hub/Controller/WebsitesController.php]: `return new JsonResponse($jsonWebsites);`?

File: src/Sylius/Hub/Security/InMemoryApiUserProvider.php [hunk 10]
```diff
@@ -35,7 +35,11 @@ public function __construct(Container $container)
      */
     public function loadUserByUsername($username)
     {
-        return $this->getUser($username);
+        if($this->user->getUsername() === $username) {
```
Comment (pamil) [inline src/Sylius/Hub/Security/InMemoryApiUserProvider.php]: CS

File: src/Sylius/Hub/Security/InMemoryApiUserProvider.php [hunk 11]
```diff
@@ -61,42 +65,25 @@ public function supportsClass($class)
...
-        if (!isset($this->users[strtolower($username)])) {
-            $exception = new UsernameNotFoundException(sprintf('Username "%s" does not exist.', $username));
-            $exception->setUsername($username);
-
-            throw $exception;
+        if (null !== $this->user && $this->user->getUsername() === strtolower($user->getUsername())) {
+            throw new \LogicException('User with this username already exists.');
         }
 
-        return $this->users[strtolower($username)];
```
Comment (pamil) [inline src/Sylius/Hub/Security/InMemoryApiUserProvider.php]: What's the reason of this change?

File: src/Sylius/Hub/Controller/StatisticsController.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Sylius\Hub\Controller;
+
+use Sylius\Hub\Repository\WebsiteRepositoryInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class StatisticsController
```
Comment (NeverResponse) [inline src/Sylius/Hub/Controller/StatisticsController.php]: `WebsitesStatisticsController`? Or it could be merged together with `WebsiteController`, though i think statistics controller alone could become pretty large quite fast, so dunno if there's any sense in combining them.

File: src/Sylius/Hub/Controller/StatisticsController.php (hunk 8 above)
Comment (NeverResponse) [inline src/Sylius/Hub/Controller/StatisticsController.php]: Docblock missing

File: src/Sylius/Hub/Controller/StatisticsController.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+    public function __construct(WebsiteRepositoryInterface $websiteRepository)
+    {
+        $this->websiteRepository = $websiteRepository;
+    }
+
+    /**
+     * @return JsonResponse
+     */
+    public function getTotalAction()
```
Comment (NeverResponse) [inline src/Sylius/Hub/Controller/StatisticsController.php]: `countDistinctWebsitesAction` or option w/o the `Websites` part if the whole controller's name will change.

File: spec/Sylius/Hub/Entity/WebsiteSpec.php [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    function its_platform_is_mutable()
+    {
+        $this->setPlatform('Mozilla/4.78 [en] (Win98; U)');
+        $this->getPlatform()->shouldReturn('Mozilla/4.78 [en] (Win98; U)');
+    }
+
+    function its_last_contact_date_is_mutable()
+    {
+        $this->setLastContactDate(new \DateTime('2000-01-01'));
+        $this->getLastContactDate()->shouldReturnAnInstanceOf(\DateTime::class);
```
Comment (tuka217) [inline spec/Sylius/Hub/Entity/WebsiteSpec.php]: 👍

File: spec/Sylius/Hub/Security/TokenAuthenticatorSpec.php (hunk 5 above)
Comment (tuka217) [inline spec/Sylius/Hub/Security/TokenAuthenticatorSpec.php]: 👍

File: src/Sylius/Hub/Controller/WebsitesController.php (hunk 9 above)
Comment (tuka217) [inline src/Sylius/Hub/Controller/WebsitesController.php]: Cannot be, because then we have serialization done in 42 line and here

File: src/Sylius/Hub/Security/InMemoryApiUserProvider.php (hunk 11 above)
Comment (tuka217) [inline src/Sylius/Hub/Security/InMemoryApiUserProvider.php]: We have (and want to have) only one user in Hub, so the array of users is now not needed.

File: tests/Sylius/Hub/Responses/Expected/websites/authentication_required_response.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+  "message":"Authentication Required"
```
Comment (NeverResponse) [inline tests/Sylius/Hub/Responses/Expected/websites/authentication_required_response.json]: missing space after `:`

File: src/Sylius/Hub/Controller/PackagistController.php [hunk 16]
```diff
@@ -1,37 +1,38 @@
...
+use Sylius\Hub\Provider\SyliusVersionProviderInterface;
 use Symfony\Component\HttpFoundation\JsonResponse;
-use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\HttpKernel\Exception\HttpException;
 
 /**
  * @author Anna Walasek <anna.walasek@lakion.com>
  */
-final class PackagistController extends Controller
```
Comment (NeverResponse) [inline src/Sylius/Hub/Controller/PackagistController.php]: 👍

File: src/Sylius/Hub/Security/InMemoryApiUserProvider.php (hunk 11 above)
Comment (pamil) [inline src/Sylius/Hub/Security/InMemoryApiUserProvider.php]: Yet we still have some conditionals like `$this->user->getUsername() === $username` in `loadUserByUsername` so it causes a lot of confusion for project newcomers. Anyway, if you decide to stay with this approach, some class-level comment about that would be appreciated.

File: src/Sylius/Hub/Controller/StatisticsController.php (hunk 12 above)
Comment (tuka217) [inline src/Sylius/Hub/Controller/StatisticsController.php]: I think, that in the future the StastisticsController will be large. But we should also create the code without taking into consideration the future.

File: src/Sylius/Hub/Controller/WebsitesController.php (hunk 9 above)
Comment (tuka217) [inline src/Sylius/Hub/Controller/WebsitesController.php]: Then you set json data as json data of the request (and it should be raw data set as json request). So, should stay the response (as I want to have the serialization from 42 line).

File: src/Sylius/Hub/Security/InMemoryApiUserProvider.php (hunk 11 above)
Comment (tuka217) [inline src/Sylius/Hub/Security/InMemoryApiUserProvider.php]: Truly I can not implement the loadUserByUsername method at all, because currently it is not needed (but I need to implements UserProviderInterface, which contains this method). So, I'm not sure which solution is better. Stay with array of users and use loadUserByUsername or throw in loadUserName an exception and have only one user?

File: src/Sylius/Hub/Controller/WebsitesController.php (hunk 9 above)
Comment (pamil) [inline src/Sylius/Hub/Controller/WebsitesController.php]: All right, `return JsonResponse::fromJsonString($jsonWebsites);` should do the job then 🎉

File: src/Sylius/Hub/Controller/StatisticsController.php (hunk 12 above)
Comment (tuka217) [inline src/Sylius/Hub/Controller/StatisticsController.php]: On the other hand, the `WebsitesController` is more like crud and the `StatisticsController` is a different concept.

**Comment by NeverResponse** [comment]
Thanks Ania!
